cis194/week2/daniel-mc/LogAnalysis.hs

```diff
@@ -0,0 +1,65 @@
+{-# OPTIONS_GHC -Wall #-}
+module LogAnalysis where
+
+import           Log
+
+-- Exercise 1
+
+safeRead :: Read a => String -> Maybe a
+safeRead x = case reads x of
+  [(r, "")] -> Just r
+  _         -> Nothing
+
+parseMessageType :: [String] -> Maybe (MessageType, [String])
+parseMessageType ("E" : (safeRead -> Just level) : rest) = Just (Error level, rest)
+parseMessageType ("W" : rest) = Just (Warning, rest)
+parseMessageType ("I" : rest) = Just (Info, rest)
+parseMessageType _ = Nothing
+
+parseTsMsg :: [String] -> Maybe (TimeStamp, String)
+parseTsMsg ((safeRead -> Just ts) : rest) = Just (ts, unwords rest)
+parseTsMsg _                              = Nothing
+
+parseMessageArray :: [String] -> LogMessage
+parseMessageArray (parseMessageType -> Just (msgType, parseTsMsg -> Just (ts, msg))) =
+  LogMessage msgType ts msg
+parseMessageArray msg = Unknown (unwords msg)
+
+parseMessage :: String -> LogMessage
+parseMessage = parseMessageArray . words
+
+parse :: String -> [LogMessage]
+parse x = map parseMessage (lines x)
```

Comment:
```Haskell 
parse x = map parseMessage (lines x)
```
is the same as
```Haskell 
parse x = map parseMessage $ lines x
```
is the same as 
```Haskell 
parse x = map parseMessage . lines $ x
```
is the same as 
```Haskell 
parse = map parseMessage . lines
```
The last one is preferred form in Haskell